Imagecaption

lisabug opened the pull request:
I implemented a basic image caption example by combining VGG16 and LSTM.
But on MSCOCO dataset, the training speed is about almost 15 samples/s, which is definitely slow
compared to neuraltalk2 by torch.
I'm glad if you could help me to improve the training speed.:)

tqchen commented:
cc @pluskid 

- What is the batch size used in the dataset?
- Can you fixed the convolution stage and test the speed on the LSTM side?

I think it is likely due to the usage of end-to-end training with multiple GPU. But I am not very sure aboutt it. Current LSTM interface is usually faster with CPu

piiswrong commented on example/image-caption/train.py:
```diff
@@ -0,0 +1,123 @@
+    import argparse
+    import os
+
+    os.environ["MXNET_EXEC_INPLACE_GRAD_SUM_CAP"]= '100'
```
this actually slows down things by a lot. The warnings are "harmless", try testing speed without this line.

lisabug commented:
@tqchen Thanks for your advice.
* I trained it with batch size of 16 on single GPU Titan Black.
* I'll test only LSTM side later.
* yes, training with GPU is faster than CPU, though my implementation is still not fast.

@piiswrong 
Firstly I trained it without changing environment variable, speed was almost same.

pluskid commented:
@antinucleon Do we have the same speed as torch when running on PTB? One potential factor for slowness that @tqchen noticed is that we do embedding separately for each time index, as oppose to doing them all at once and splitting. 

@lisabug What is the reference speed of neuraltalk2 running on the same card? Just to get a feeling of how much slower we are.

lisabug commented:
@tqchen @pluskid
I'm sorry I missed a important reason why is neuraltalk2 is fast.
In neuraltalk2, input consists one images and five correspond sentences. CNN only forward one time and a feature expand layer repeatedly expands features to number of correspond sentence, which causes that even batch size is m but get about speed of five times fast.
In my implementation, I divide (image, 5 captions) pair to (image, caption) pair in order to handle different length buckets, and it makes my implementation slow.
Sorry for myself's miss, I'll rewrite it and make a another PR. Thanks :)

Ps, some test data on single Titan Black
* Fixed cnn, without changing environment variable and LSTM length 16, speed is about 500 samples/s.
* Fixed cnn, with changing environment variable and LSTM length 16, speed is about 440 samples/s. The environment variable influences speed definitely. @piiswrong 
* NeuralTalk2 speed, with feature expands, is about 14.5 (image, 5 * captions) pairs/s.

Thanks for all your advising.

tqchen commented:
We can possibly use two different modules for LSTM and cnn, as what @pluskid did in #1787

tqchen commented:
You can also change parallelism level on each GPU on such setting, see http://mxnet.readthedocs.org/en/latest/env_var.html

MXNET_EXEC_NUM_TEMP and MXNET_GPU_WORKER_NTHREADS are relevant env vars

lisabug commented:
@tqchen Thanks a lot. It's a good idea to use different modules, just like NeuralTalk2 does.
Since fine-training CNN takes most of time of overall training, I intend to use single bucket instead of multi buckets for accelerating training. Pipeline is similar to NeuralTalk2.

lisabug commented:
Sorry for committing a simple version with bug.
I met the problem with:
```
[04:09:18] /home/lu/Source/mxnet/dmlc-core/include/dmlc/logging.h:235: [04:09:18] src/operator/./embedding-inl.h:85: Check failed: (req[embedding::kData]) == (kNullOp) Embedding layer doesn't support calculate data gradient
[04:09:18] /home/lu/Source/mxnet/dmlc-core/include/dmlc/logging.h:235: [04:09:18] src/engine/./threaded_engine.h:295: [04:09:18] src/operator/./embedding-inl.h:85: Check failed: (req[embedding::kData]) == (kNullOp) Embedding layer doesn't support calculate data gradient
```
Training performs well at first iteration but died at the next..
I have checked my code and I still don't know where the problem is or if it's a bug.
Could anyone help me and give me some advise? Thank you.

pluskid commented:
@lisabug I'm guessing: are you using a `SequentialModule` to chain a convolution module with an RNN module? That assertion would not happen normally. It might happen if, for example, you are using a `Module` and turned on `inputs_need_grad` bit.

lisabug commented:
@pluskid actually in order to solve the problem, I just used executor instead of module. Bp doesn't go to cnn yet. 
Compared to example lstm_bucketing, this lstm part just have a additional image feature input at timestep 0, and other parts are almost same. It is weirded.

pluskid commented:
So do you mean the error occurs at the very beginning or do you mean that for the first iteration it trains fine, and then the second iteration the error occurred?

lisabug commented:
@pluskid Yes, exactly as you talk. It is possible that something wrong with input data?

piiswrong commented:
@lisabug Any updates on this? Lets merge in a working version first

lisabug commented:
@piiswrong Not yet but I'll follow it soon, maybe this weekend,

sxjscience commented:
How about to change the example image?https://github.com/lisabug/mxnet/blob/31c995724d971379b60f3831b47e8db52cb45a11/example/image-caption/test.jpg

tqchen commented:
@lisabug any updates on this?

winstywang commented:
Close for now, since the authors cannot find the files....
